src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,260 @@
+-- @
+-- scc 'empty'      == 'empty'
+-- scc ('vertex' x) == 'vertex' (NonEmpty.'NonEmpty.vertex' x)
+-- @
+scc :: (Ord a) => AM.AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
+scc = AAM . AM.scc
+
+-- | Compute the /topological sort/ of a graph.
+-- 
+-- @
+-- topSort (1)             == [1]
+-- topSort (1 * 2 + 3 * 1) == [3,1,2]
+-- @
+topSort :: (Ord a) => AdjacencyMap a -> [a]
+topSort (AAM am) = Typed.topSort (Typed.fromAdjacencyMap am)
+
+-- | Compute the /Cartesian product/ of graphs.
+-- Complexity: /O(s1 * s2)/ time, memory and size, where /s1/ and /s2/
+-- are the sizes of the given graphs.
+--
+-- @
+-- edgeList $ box (1 * 2) (3 * 4) == [((1,3),(1,4))
+--                                   ,((1,3),(2,3))
+--                                   ,((1,4),(2,4))
+--                                   ,((2,3),(2,4))]
+-- edgeList $ box (1 + 2) (3 + 4) == [(1,3),(1,4),(2,3),(2,4)]
+-- @
+box :: (Ord a, Ord b) => AdjacencyMap a -> AdjacencyMap b -> AdjacencyMap (a, b)
+box (AAM x) (AAM y) = AAM $ AM.overlay (fO $ xy y x) (fO $ yx x y)
+  where
+    xy q = AM.gmap (\x -> AM.gmap (x, ) q)
+    yx p = AM.gmap (\y -> AM.gmap (, y) p)
+    fO =
+      Map.foldrWithKey (\k -> AM.overlay . Set.foldr AM.overlay k) AM.empty .
+      AM.adjacencyMap
+
+-- | This is a signature for a 'Strict Partial Order'.
+-- A strict partial order is a binary relation 'R' that has three
+-- axioms, namely, irreflexive, transitive and asymmetric.
+--   > a 'R' a == False                    (Irreflexive)
+--   > a 'R' b and b 'R' c => a 'R' c      (Transitive)
+--   > a 'R' b == True => b 'R' a == False (Asymmetric)
+-- Some examples of a Strict Partial Order are (<) and (>).
+type PartialOrder a = a -> a -> Bool
+
+-- | Constructs an acyclic graph from any graph based on
+-- a strict partial order to produce an acyclic graph.
+-- The partial order defines the valid set of edges.
+-- For example, if the partial order is (<) then for any two
+-- vertices x and y (x > y), the only possible edge is (y, x).
+-- This will guarantee the production of an acyclic graph as no
+-- back edges are nor possible.
+--
+-- For example,
+-- 'fromGraph (<) (1 * 2 + 2 * 1) == 1 * 2' because
+-- '1 < 2 == True' and hence the edge is allowed.
+-- '2 < 1 == False' and hence the edge is filtered out.
+--
+-- @
+-- fromGraph (<) (2 * 1)         == 'empty'
+-- fromGraph (<) (1 * 2)         == 1 * 2
+-- fromGraph (<) (1 * 2 + 2 * 1) == 1 * 2
+-- @
+fromGraph :: Ord a => PartialOrder a -> Graph a -> AdjacencyMap a
+fromGraph o = AAM . aMF . foldg AM.empty AM.vertex AM.overlay AM.connect
+  where
+    aMF = AM.AM . Map.mapWithKey (\k -> Set.filter (k `o`)) . AM.adjacencyMap
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 
+-- 'AdjacencyMap' for more details.
+instance (Ord a, Num a) => Num (AdjacencyMap a) where
+  fromInteger = AAM . fromInteger
+  (AAM (AM.AM x)) + (AAM (AM.AM y)) =
+    AAM . AM.AM $
+    Map.unionWithKey (\k p q -> Set.filter (> k) $ Set.union p q) x y
```

Comment:
Should I mention that by default the partial order is (<) in the instance 'Num'?
ie. an edge (x, y) is possible if and only if 'x < y == True'